### PR TITLE
Add function management APIs to the runtime

### PR DESCRIPTION
**Pull request by ahmelsayed**
This change adds the following APIs:

## Function management APIs
* `GET /admin/functions` - List available functions
* `GET /admin/functions/{name}` - Get a specific function - `FunctionsController`
* `PUT /admin/functions/{name}` - Create or update an existing function - `FunctionsController`
* `DELETE /admin/functions/{name}` - Delete a function - `FunctionsController`

## File management APIs (VFS)
* `GET /admin/vfs/*` - Get a file or directory content - `VirtualFileSystem`
* `PUT /admin/vfs/*` - Update file or directory - `VirtualFileSystem`
* `DELETE /admin/vfs/*`  - Delete a file or directory - `VirtualFileSystem`
* `GET /admin/zip/*` - Download a zip archive - `ZipFileSystem`

The VFS implementation is mostly a copy of the one in kudu, but modified to work with ASP.NET Core and on Linux. However there is a difference in behavior between Unix vs Windows for vfs.
In Windows, the file system doesn’t have a root (at least not in win32) so the root of your virtual filesystem is your home. (%HOME%) or scriptRoot if HOME is not defined, and is a special path that looks like `/admin/vfs/SystemDrive`, this maps to the system drive.
In Linux the root is simply just / which is the root of the file system.

This PR also:

* Renames and splits `AdminController`  --> `HostController` for script host general APIs both runtime and management, `FunctionsController` for function specific APIs, both runtime and management.
* Removes check for host running before any `/admin` API call. Basically whoever implements an Admin API has to expect the host to not be running. If it's a management API, then it should work. If it's a runtime api, it should fail with a proper error explaining why the host is not running.

//cc @davidebbo @btardif  for FYI

### CODE REVIEW

**Comment by fabiocav** [comment]
@ahmelsayed apologies for the delay on this... could you rebase this on the latest before a review?

**Comment by ahmelsayed** [comment]
will do

I figured I should explain a bit about the PR and file structure since there are a lot of new files. 

```zsh
src/WebJobs.Script.WebHost/Management
|- IWebFunctionsManager.cs # interface implemented in WebFunctionsManager.cs below
|- VirtualFileSystem.cs # implementation of /vfs/ apis
|- VistualFileSystemBase.cs # shared code between VFS and ZIP file systems
|- WebFunctionsManager.cs # implementation that drives the management APIs in FunctionsController
|- ZipFileSystem.cs # allows uploading zip files, and downloading directories as zip files.

src/WebJobs.Script.WebHost/Extensions 
|- .. # Helper classes

src/WebJobs.Script.WebHost/Helpers
|- FileSystemHelpers.cs # You mentioned that you use System.IO.Abstraction (or want to use?) I couldn't find it. 

src/WebJobs.Script.WebHost/Middleware
|- # I moved VFS from kudu as a middleware rather than a controller because all the code is expecting to get an HttpRequestMessage, and produce an HttpResponseMessage. It doesn't map itself too easily to IActionResult from what I have seen. Though I'm open to feedback if you think differently. 

src/WebJobs.Script.WebHost/Models
```

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs [hunk 1]
```diff
@@ -29,76 +31,22 @@ namespace Microsoft.Azure.WebJobs.Script.WebHost.Controllers
     /// Controller responsible for handling all administrative requests, for
     /// example enqueueing function invocations, etc.
     /// </summary>
-    public class AdminController : Controller
+    public class HostController : Controller
     {
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Controllers/HostController.cs]: Update description?

**Comment by ahmelsayed** [comment]
oh I thought I rebased this, but it looks like dev moved again :)

File: src/WebJobs.Script.WebHost/Middleware/VirtualFileSystemMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                    response.Dispose();
+                }
+
+                await context.Response.WriteAsync(e.Message);
+            }
+        }
+
+        private async Task<bool> AuthenticateAndAuthorize(HttpContext context)
+        {
+            var authorizationEvaluator = context.RequestServices.GetRequiredService<IAuthorizationPolicyProvider>();
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Middleware/VirtualFileSystemMiddleware.cs]: nit: change variable name to `authorizationPolicyProvider` or `prolicyProvider`

File: src/WebJobs.Script.WebHost/Models/FunctionMetadataResponse.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace WebJobs.Script.Management.Models
+{
+    public class FunctionMetadataResponse
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Models/FunctionMetadataResponse.cs]: How problematic would it be if we changed the response payloads to use the ApiModel approach used with other management APIs? That model handles things like ref links (self, hrefs, etc.) and other things to keep the way resources are exposed consistent

File: src/WebJobs.Script.WebHost/WebHostResolver.cs [hunk 4]
```diff
@@ -199,23 +202,25 @@ internal static ScriptHostConfiguration CreateScriptHostConfiguration(WebHostSet
             return scriptHostConfig;
         }
 
-        private static void InitializeFileSystem()
+        private static void InitializeFileSystem(WebHostSettings settings)
         {
             if (ScriptSettingsManager.Instance.IsAzureEnvironment)
             {
+                // When running on Azure, we kick this off on the background
+                // Q: why?
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/WebHostResolver.cs]: Not sure if the logic here was "run on background when running on Azure" or "run on background if there's work to do", which only happened on Azure. The intention here is not really clear.

File: src/WebJobs.Script.WebHost/WebHostSettings.cs [hunk 5]
```diff
@@ -24,6 +23,8 @@ public class WebHostSettings
 
         public string SecretsPath { get; set; }
 
+        public string TestDataPath { get; set; }
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/WebHostSettings.cs]: Can you add a description? What is this for?

File: src/WebJobs.Script.WebHost/WebJobsServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -82,6 +85,10 @@ public static IServiceProvider AddWebJobsScriptHost(this IServiceCollection serv
             // The services below need to be scoped to a pseudo-tenant (warm/specialized environment)
             builder.Register<WebScriptHostManager>(c => c.Resolve<WebHostResolver>().GetWebScriptHostManager()).ExternallyOwned();
             builder.Register<ISecretManager>(c => c.Resolve<WebHostResolver>().GetSecretManager()).ExternallyOwned();
+            builder.RegisterType<WebFunctionsManager>().As<IWebFunctionsManager>().SingleInstance();
+            builder.RegisterType<VirtualFileSystem>();
+            builder.RegisterType<ZipFileSystem>();
+            builder.RegisterType<VirtualFileSystemMiddleware>();
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/WebJobsServiceCollectionExtensions.cs]: Was the middleware registration here required? THe call to UseMiddleware should be sufficient

File: src/WebJobs.Script.WebHost/Extensions/TaskExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Extensions
+{
+    public static class TaskExtensions
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Extensions/TaskExtensions.cs]: Class name confused me at first. This is really an IEnumerable of Task extensions class

File: src/WebJobs.Script.WebHost/Helpers/DelegatingStream.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+{
+    public abstract class DelegatingStream : Stream
+    {
+        private Stream _innerStream;
+
+        protected DelegatingStream(Stream innerStream)
+        {
+            if (innerStream == null)
+            {
+                throw new ArgumentNullException("innerStream");
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Helpers/DelegatingStream.cs]: nit: use `nameof` for parameter name and throw expression for the `null` validation.

File: src/WebJobs.Script.WebHost/Helpers/DelegatingStream.cs [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+using System;
+using System.IO;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Helpers
+{
+    public abstract class DelegatingStream : Stream
+    {
+        private Stream _innerStream;
+
+        protected DelegatingStream(Stream innerStream)
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Helpers/DelegatingStream.cs]: Why is this needed?

File: src/WebJobs.Script.WebHost/Helpers/FileSystemHelpers.cs [hunk 10]
```diff
@@ -0,0 +1,114 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Helpers
+{
+    public static class FileSystemHelpers
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Helpers/FileSystemHelpers.cs]: There's a `FileUtility` class already in the project. Some things here seem to overlap, but we should either move what is here there or vice-versa, but we shouldn't have both.

File: src/WebJobs.Script.WebHost/Extensions/HttpRequestExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Extensions;
+
+namespace WebJobs.Script.WebHost.Extensions
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Extensions/HttpRequestExtensions.cs]: We have an `HttpRequestExtensions` class in Script, would be good to just move this there.

File: src/WebJobs.Script.WebHost/Middleware/ScriptHostCheckMiddleware.cs [hunk 12]
```diff
@@ -1,9 +1,6 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the MIT License. See License.txt in the project root for license information.
```
Comment (fabiocav) [inline src/WebJobs.Script.WebHost/Middleware/ScriptHostCheckMiddleware.cs]: Is the change here required? Just removing unused usings? Or was there something that got lost in a rebase?

**Comment by fabiocav** [comment]
@ahmelsayed let me know if you need any more information on the feedback provided. Would be good to have these changes merged in soon.

File: src/WebJobs.Script.WebHost/WebHostResolver.cs (hunk 4 above)
Comment (ahmelsayed) [reply src/WebJobs.Script.WebHost/WebHostResolver.cs]: it just seemed odd to me that this path is running in a task. I wasn't sure why so I put the comment. Still not really sure why it was done like this in the first place :)

File: src/WebJobs.Script.WebHost/Middleware/VirtualFileSystemMiddleware.cs (hunk 2 above)
Comment (ahmelsayed) [reply src/WebJobs.Script.WebHost/Middleware/VirtualFileSystemMiddleware.cs]: done

File: src/WebJobs.Script.WebHost/Controllers/HostController.cs (hunk 1 above)
Comment (ahmelsayed) [reply src/WebJobs.Script.WebHost/Controllers/HostController.cs]: done

File: src/WebJobs.Script.WebHost/Helpers/DelegatingStream.cs (hunk 9 above)
Comment (ahmelsayed) [reply src/WebJobs.Script.WebHost/Helpers/DelegatingStream.cs]: some random C# quirk with inheriting from Stream class I guess. It's only used in the ZipArchive scenario and I'm removing that since it doesn't make sense in the runtime.

File: src/WebJobs.Script.WebHost/Middleware/ScriptHostCheckMiddleware.cs (hunk 12 above)
Comment (ahmelsayed) [reply src/WebJobs.Script.WebHost/Middleware/ScriptHostCheckMiddleware.cs]: obsolete since Mathew replaced this file with another implementation.

File: src/WebJobs.Script.WebHost/Extensions/HttpRequestExtensions.cs (hunk 11 above)
Comment (ahmelsayed) [reply src/WebJobs.Script.WebHost/Extensions/HttpRequestExtensions.cs]: done

File: src/WebJobs.Script.WebHost/Helpers/DelegatingStream.cs (hunk 8 above)
Comment (ahmelsayed) [reply src/WebJobs.Script.WebHost/Helpers/DelegatingStream.cs]: removed file

File: src/WebJobs.Script.WebHost/Extensions/TaskExtensions.cs (hunk 7 above)
Comment (ahmelsayed) [reply src/WebJobs.Script.WebHost/Extensions/TaskExtensions.cs]: done

File: src/WebJobs.Script.WebHost/WebJobsServiceCollectionExtensions.cs (hunk 6 above)
Comment (ahmelsayed) [reply src/WebJobs.Script.WebHost/WebJobsServiceCollectionExtensions.cs]: done

File: src/WebJobs.Script.WebHost/WebHostSettings.cs (hunk 5 above)
Comment (ahmelsayed) [reply src/WebJobs.Script.WebHost/WebHostSettings.cs]: done

File: src/WebJobs.Script.WebHost/Models/FunctionMetadataResponse.cs (hunk 3 above)
Comment (ahmelsayed) [reply src/WebJobs.Script.WebHost/Models/FunctionMetadataResponse.cs]: It's possible. It'll just require a UX change to handle both the kudu model and this new model. 

I was thinking that eventually we would get rid of `FunctionMetadataResponse` object since it's just a [map](https://github.com/Azure/azure-webjobs-sdk-script/blob/48b164b08b97b3401712469089128e628eba2f1b/src/WebJobs.Script.WebHost/Extensions/FunctionMetadataExtensions.cs#L29) on top of `FunctionMetadata`. also kudu uses `name_name` instead of `camelCasing` like the runtime. I wanted to change the model but would rather just do it once so that the UX expects only one or the other.

so do you think it's better to keep both `FunctionMetadata` and `FunctionMetadataResponse`?

File: src/WebJobs.Script.WebHost/Extensions/FunctionMetadataExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+            var functionMetadataFilePath = Path.Combine(functionPath, ScriptConstants.FunctionMetadataFileName);
+            var baseUrl = request != null
+                ? $"{request.Scheme}://{request.Host}"
+                : "https://localhost/";
+
+            var response = new FunctionMetadataResponse
+            {
+                Name = functionMetadata.Name,
+
+                // Q: can functionMetadata.ScriptFile be null or empty?
```
Comment (ahmelsayed) [inline src/WebJobs.Script.WebHost/Extensions/FunctionMetadataExtensions.cs]: @fabiocav do you know the answer to this?

**Comment by ahmelsayed** [comment]
@fabiocav  regarding tests, most of the VFS tests in kudu rely on System.IO.Abstraction mocks which the WebHost project doesn't use. Would you be okay with me switching `FileUtility` to use `IFileSystem` instead of directly using `System.IO.*`?

also sorry for the delay in addressing your comments :) I was swamped by so many different things the past couple of months.

**Comment by ahmelsayed** [comment]
I updated `FileUtility` to use `System.IO.Abstraction` and added kudu tests for vfs [here](https://github.com/Azure/azure-functions-host/blob/6dca084f61567ae962d8a83b5c05434825ff247d/test/WebJobs.Script.Tests/Managment/VirtualFileSystemFacts.cs)

**Comment by ahmelsayed** [comment]
I just notices that all tests are now passing on dev. I just did a rebase and will take a look at any failed tests

**Comment by ahmelsayed** [comment]
okay, all tests should be passing now, but few things:

- This commit 0518ba8df0d0b5d2ac338f78f273a886bbcc7a31 clears up the `Mock<IFileSystem>` that the vfs test cases use. This will need to be reconsidered if you ever plan on parallel test cases. 

- This commit 627e71660cda6649b41d2eca0bb8452a43259e4d however changes an existing test case. it seems that the test case is trying to handle the secret file being locked and verifying that the correct exception is raised. However, it looks like the original `FileUtility.ReadAsync` was just using a `StreamReader` which locks the file by default. 

https://github.com/Azure/azure-functions-host/blob/8cf2debe0f4896f4433d96b828b6586a06554293/src/WebJobs.Script/Extensions/FileUtility.cs#L64-L76


I updated `FileUtility.ReadAsync` to only grab a read lock to the file. 

https://github.com/Azure/azure-functions-host/blob/627e71660cda6649b41d2eca0bb8452a43259e4d/src/WebJobs.Script/Extensions/FileUtility.cs#L75-L88

I can either revert that without changing the test, or changing the test to allow `FileShare.None` when locking the file.

**Comment by ahmelsayed** [comment]
yay! 

/cc @christopheranderson

**Comment by brettsam** [comment]
It looks like this change adds two matching routes for `GET /admin/functions/{name}/status`.
https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script.WebHost/Controllers/HostController.cs#L79
https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script.WebHost/Controllers/FunctionsController.cs#L116

The e2e tests I'm working on caught it as I rebased. I'm removing the one in `HostController` locally to unblock myself.

**Comment by brettsam** [comment]
Same with `POST admin/functions/{name}` (Invoke).

**Comment by ahmelsayed** [comment]
sorry about that. It looks like one of the rebases I did brought it back and I didn't notice

**Comment by fabiocav** [comment]
Missed on the last review as well... That's the risk when those changes become large. We'll just adjust, get it fixed and be happy :)
